### PR TITLE
Performance improvements (removing Meriyah)

### PR DESCRIPTION
**Pull request by vrugtehagel**
This is a redo of https://github.com/ventojs/vento/pull/123, without the `js.ts` or `tokenizer.ts`-specific parts. It iterates the generated code to detect used variables using the same `iterateTopLevel()` function, and then prepends a bit of code to define said variables.

It does speed up Vento quite a bit, though the "render" phase (which was insanely fast before) is a little less fast (although still beating nearly every other templating language). The render phase is insignificant compared to compilation, but if a template is reused a lot, this can add up.

### CODE REVIEW

**Comment by oscarotero** [comment]
Thanks! I need time to analyze this because it's a huge change and can break existing sites.

FYI, while testing this, I just discovered a huge bug in the caching mechanism, causing the same template to be compiled several times (it was related with async functions). I just fixed it in this commit 3e7c831b29230042bbc6b81d2dfb07718e34e522 and clearly this is a big improvement, specially knowing that the Vento's compilation is not the fastest. 

I have tested this change with a real site (the lume.land website) and it works fine, with no errors.
In [this template](https://github.com/lumeland/lume.land/blob/main/_includes/templates/type.vto) it detects the following variables:
```
def, Array, child, name, info, of, options, printDef, data, env, JSON, Deno, __tmp
```

I think the slowness of render phase is caused by declaring unnecesary variables. `Array`, `JSON` or `Deno` are global variables, I think it could check if this variable exists in `globalThis` before saving them (there's [a function for that here](https://github.com/ventojs/vento/blob/main/src/environment.ts#L325)). The `of` variable should be avoided too, because it's used in the loops (`for...of`).

It would be nice if other variables like printDef (that is a function) or `def` (which is the function argument) could be avoided, although I guess it's not so easy.

**Comment by vrugtehagel** [comment]
Thanks for the quick response! But yes, _please_ take your time reviewing this. The tests may pass, but it is a huge change regardless that could very well contain some sneaky edge cases.

For the globals - initially I didn't filter them out, because you _can_ redeclare them (e.g. `const Array = [1,2,3]` is fine in JS) but I did not realize that passing a variable with such a name to a template does not overwrite the global. I mean, if I do `{{ include "tmp.vto" { Array: [1, 2, 3] } }}`, then the `Array` variable inside `tmp.vto` is still just the global `Array`, not `[1, 2, 3]`. I don't know if this is necessarily a good design decision, because it means other code can mess with how Vento builds (e.g. if a dependency of lume.land's site defines `globalThis.options`, then the template you linked no longer works as expected).

Either way, that's just how Vento works right now, so removing globals from the list of variables seems safe. With that (and reserving `__tmp`), the template you linked only generates declarations for

```
printDef, def, child, name, info, of, options
```

and it does so much more compactly with variable destructuring. So that's an improvement already.

For detecting local variables, I reckon this is not worth the complexity; we hypothetically can detect top-level declarations and filter those out, but it is a tradeoff that I'm not sure ends up being a net gain in terms of performance, and definitely won't be in terms of simplicity/maintainability.

As for avoiding `of` - this is a valid variable name. I guess it falls in the same bucket as `let` and `async` (and a few others), but `var let;for(let of of let);` is (oddly enough) not a syntax error. I can add `of` to the reserved variable list, but for people that like abbreviating their variable names I reckon it might not be so uncommon to use `of` (to stand for e.g. "output file"). But I don't feel strongly about this, so whichever way you think is best, we can go with.

As a sidenote, I realized the lookbehind for `const`, `let`, `var` and `function` was probably pointless, because almost anything that is locally declared in a template is also used. And if it is used, it will be picked up anyway. So I've simplified the regex by removing that part of the lookbehind, although I kept the `.` and `?.` lookbehind (this does help a lot).

File: src/environment.ts [hunk 1]
```diff
@@ -320,7 +324,7 @@ export class Environment {
 }
 
 function isGlobal(name: string) {
-  // @ts-ignore TS doesn't know about globalThis
+  if (name == "name") return false;
```
Comment (vrugtehagel) [inline src/environment.ts:327]: I am specifically filtering out `"name"` here; it is a global (equal to the empty string), but also a very common local variable name, so we should allow it in templates. The original use of `isGlobal()` is [here](https://github.com/ventojs/vento/blob/v1.14.0/src/environment.ts#L297-L301), and changing the behavior for `"name"` fortunately doesn't break current behavior (using `|> name` would throw an error in current Vento anyway).

**Comment by vrugtehagel** [comment]
Awesome! Maybe we can even use it for a playground on Vento's site, that would be great!

**Comment by oscarotero** [comment]
Thanks @wrapperup for testing it!
Yeah, I was also a bit concern about the code transformation, but the new approach of detecting and declaring the used variables first seems to work nicely! And now Vento 2 is dependency-free!
